extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/CodeAuthenticationMechanism.java

```diff
@@ -76,21 +77,50 @@
+        final Cookie stateCookie = context.request().getCookie(getStateCookieName(oidcTenantConfig));
+
+        // if the state cookie is available then try to complete the code flow and start a new session
+        if (stateCookie != null) {
+
+            if (!isStateCookieValid(context, stateCookie)) {
+                return Uni.createFrom().failure(new AuthenticationCompletionException());
+            }
+
+            final String code = context.request().getParam(OidcConstants.CODE_FLOW_CODE);
+            if (code != null) {
+                // start a new session by starting the code flow dance
+                Uni<TenantConfigContext> resolvedContext = resolver.resolveContext(context);
+                return resolvedContext.onItem()
+                        .transformToUni(new Function<TenantConfigContext, Uni<? extends SecurityIdentity>>() {
+                            @Override
+                            public Uni<SecurityIdentity> apply(TenantConfigContext tenantContext) {
+                                return performCodeFlow(identityProviderManager, context, tenantContext, code, stateCookie);
+                            }
+                        });
+            } else {
+                LOG.debug("State cookie is present but neither 'code' nor 'error' query parameter is returned");
+                OidcUtils.removeCookie(context, oidcTenantConfig, stateCookie.getName());
+                return Uni.createFrom().failure(new AuthenticationCompletionException());
+            }
         }
+        // return an empty identity - this will lead to a challenge redirecting the user to OpenId Connect provider
+        // unless it is detected it is a redirect from the provider in which case HTTP 401 will be returned.
+        context.put(NO_OIDC_COOKIES_AVAILABLE, Boolean.TRUE);
+        return Uni.createFrom().optional(Optional.empty());
 
-        // start a new session by starting the code flow dance
-        Uni<TenantConfigContext> resolvedContext = resolver.resolveContext(context);
-        return resolvedContext.onItem()
-                .transformToUni(new Function<TenantConfigContext, Uni<? extends SecurityIdentity>>() {
-                    @Override
-                    public Uni<SecurityIdentity> apply(TenantConfigContext tenantContext) {
-                        return performCodeFlow(identityProviderManager, context, tenantContext, code);
-                    }
-                });
+    }
 
+    private boolean isStateCookieValid(RoutingContext context, Cookie stateCookie) {
+        List<String> values = context.queryParam("state");
+        // IDP must return a 'state' query parameter and the value of the state cookie must start with this parameter's value
+        if (values.size() != 1) {
+            LOG.debug("State parameter can not be empty or multi-valued");
+            return false;
+        } else if (!stateCookie.getValue().startsWith(values.get(0))) {
```

Comment:
Isn't more secure an exact match?

Reply:
Hi @pedroigor This is not specific to this PR, I just moved some of the state check here but see on main [here](https://github.com/quarkusio/quarkus/blob/main/extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/CodeAuthenticationMechanism.java#L266), it has been like that for ever. The state cookie value has the format `state|path-to-restore-after-redirect` (the 2nd part can hold query parameters and supports the old case where the callback URI is only used different to the actual destination URI - that can probably be dropped - but it would be too sensitive right now). This is why it is `startsWith` check.
However, good you've spotted it as we should follow up with the exact match on the 1st part of the value. I'll update